modernize: createRef API, functional components with hooks

astridx opened the pull request:
Pull Request for Issue # https://github.com/mapbox/mapbox-react-examples/issues/3.

### Summary of Changes

I implemented and added all 4 examples with hooks. ~I named the new examples `v2_nameOfExample` and not `nameOfExample-with-hooks` as suggested, because I find it clearer that way.~

I would be happy if you like this.

astridx committed:
basic_v2

astridx committed:
basci_v2

astridx committed:
basic and tooltip

astridx committed:
wip

astridx committed:
wip

astridx committed:
wip

astridx committed:
overlay

astridx committed:
wip

astridx committed:
overlay and wip redux

astridx committed:
redux

astridx committed:
fof

tristen commented:
👋 @astridx this is awesome! What do you think about updating the existing examples instead of creating `v2_*` versions?

astridx committed:
docu and overwrite old versions

astridx committed:
readme

astridx committed:
no v2

astridx commented:
@tristen Thanks for your quick feedback

I think the original versions should be overwritten. I had scruples about just doing it like that. But: Someone who reads here is building something new and needs current information. 

That's why I deleted original directories in this PR and renamed my new directories.
Is that what you mean?

If these PR were merged, it would be good if the documentation https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/ was updated.

I have included a suggestion in the Basic Readme.:
https://github.com/mapbox/mapbox-react-examples/blob/c31fe8119e35293a902a83e0ecdd524b80042404/basic/README.md

tristen commented:
> If these PR were merged, it would be good if the documentation [docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/) was updated.
> 
> I have included a suggestion in the Basic Readme.:
> [`c31fe81`/basic/README.md](https://github.com/mapbox/mapbox-react-examples/blob/c31fe8119e35293a902a83e0ecdd524b80042404/basic/README.md)

👍 I can make sure that's updated! 

@astridx your PR adds a lot of boilerplate code from create-react-app. I'd like to keep the code examples minimal so the code is well understood. Thoughts on removing additions like yarn, the testing scaffold & service workers?

astridx committed:
tidy up basic and tooltip

astridx committed:
tidy up data-overlay and data-overlay-redux

astridx committed:
tidy up public folder

astridx commented:
@tristen  You're right. The unnecessary files and texts confuse others. I've cleaned up something. Can you find anything else I could take out?

tristen commented on basic/public/index.html:
```diff
@@ -1,14 +1,44 @@
+    <link href='https://api.tiles.mapbox.com/mapbox-gl-js/v1.11.1/mapbox-gl.css' rel='stylesheet' />
+    <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <meta name="theme-color" content="#000000" />
+    <meta
+      name="description"
+      content="Web site created using create-react-app"
+    />
+    <link rel="apple-touch-icon" href="%PUBLIC_URL%/logo192.png" />
+    <!--
+      manifest.json provides metadata used when your web app is installed on a
+      user's mobile device or desktop. See https://developers.google.com/web/fundamentals/web-app-manifest/
+    -->
+    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
+    <!--
+      Notice the use of %PUBLIC_URL% in the tags above.
+      It will be replaced with the URL of the `public` folder during the build.
+      Only files inside the `public` folder can be referenced from the HTML.
+
+      Unlike "/favicon.ico" or "favicon.ico", "%PUBLIC_URL%/favicon.ico" will
+      work correctly both with client-side routing and a non-root public URL.
+      Learn how to configure a non-root public URL by running `npm run build`.
+    -->
```
Can we drop these lines from each index?

tristen commented on basic/README.md:
```diff
@@ -18,3 +18,251 @@ function.
+
+#### The HTML page
+
+Open the `index.html` file in the folder `basic-with-hooks`. This file was created by [Create React App](https://github.com/facebook/create-react-app). The line has been added for our example:
```
Looks like references to `basic-with-hooks` should change to `basic`?

tristen commented on basic/README.md:
```diff
@@ -18,3 +18,251 @@ function.
+Documentation
+---
+
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
```
If we're stripping out most of the bootstrapping from create react app, Want to drop this line?

tristen commented on basic/README.md:
```diff
@@ -1,9 +1,9 @@
-Basic example
+Basic example using React Hooks
```
Consider removing "using React Hooks" from all the readme's? If these examples are just being replaced with versions using hooks I don't think the extra messaging for that is necessary

tristen commented on basic/README.md:
```diff
@@ -18,3 +18,251 @@ function.
+
+##### index.js
+
+Open the `index.js` file in the folder `basic-with-hooks/src`. This file was created by [Create React App](https://github.com/facebook/create-react-app).
```
Remove references in this doc to Create React App?

tristen commented on basic/README.md:
````diff
@@ -18,3 +18,251 @@ function.
+
+index.js
+
+```
````
```suggestion
```js
```

To add code formatting

tristen commented on basic/README.md:
````diff
@@ -18,3 +18,251 @@ function.
+
+__app.js__
+
+```
````
```suggestion
```js
```

tristen commented on basic/README.md:
````diff
@@ -18,3 +18,251 @@ function.
+
+__Map.js__
+
+```
````
```suggestion
```js
```

tristen commented on basic/README.md:
````diff
@@ -18,3 +18,251 @@ function.
+
+1. We are the first to import everything that is necessary.
+
+```
````
```suggestion
```js
```

tristen commented on basic/README.md:
````diff
@@ -18,3 +18,251 @@ function.
+
+3. Then we create the function component `Map` and export it.
+
+```
````
```suggestion
```js
```

tristen commented on basic/README.md:
````diff
@@ -18,3 +18,251 @@ function.
+
+__Map.css__
+
+```
````
```suggestion
```css
```

tristen reviewed (changes requested):
Thanks for removing much of the scaffolding from create-react-app! Left some quick thoughts for a couple of other things to change/remove

tristen commented on basic/README.md:
```diff
@@ -18,3 +18,251 @@ function.
+For more information see [Use Mapbox GL JS in a React app](https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/) 
+
+
+Documentation
```
Can this documentation actually be moved into a ticket? I can followup with the documentation team at Mapbox to make these changes!

astridx replied:
What do you mean with "Can this documentation actually be moved into a ticket?". I think this page have to be updated if we change the examples: https://docs.mapbox.com/help/tutorials/use-mapbox-gl-js-with-react/

tristen replied:
@astridx I agree! I just think the text for this should not be written in `basic/README.md`. I'm suggesting we pull this out into a [new ticket](https://github.com/mapbox/mapbox-react-examples/issues/new) with a title that says "Update the documentation with this text to account for new changes"?

astridx committed:
tidy up public index

astridx committed:
tidy up readmes

astridx commented:
I cleaned up read.me and index.php further. Is this OK?

astridx committed:
remove documentation from basic example

astridx commented:
@tristen Please see: https://github.com/mapbox/mapbox-react-examples/issues/12

tristen commented on basic/src/Map.js:
```diff
@@ -0,0 +1,49 @@
+    // Add navigation control (the +/- zoom buttons)
+    map.addControl(new mapboxgl.NavigationControl(), "top-right");
+
+    map.on("moveend", () => {
```
Can we keep using the `"move"` event here?

tristen commented on react-tooltip/src/components/Tooltip.js:
```diff
@@ -0,0 +1,17 @@
+
+  return (
+    <div id={`tooltip-${id}`}>
+      <h3>Informations</h3>
```
Can we drop this heading?

tristen commented on react-tooltip/src/components/Tooltip.js:
```diff
@@ -0,0 +1,17 @@
+  return (
+    <div id={`tooltip-${id}`}>
+      <h3>Informations</h3>
+      <strong>Layer:</strong> {feature.layer["source-layer"]}
```
```suggestion
      <strong>Source layer:</strong> {feature.layer["source-layer"]}
```

tristen commented on react-tooltip/src/components/Tooltip.js:
```diff
@@ -0,0 +1,17 @@
+      <h3>Informations</h3>
+      <strong>Layer:</strong> {feature.layer["source-layer"]}
+      <br />
+      <strong>ID:</strong> {feature.layer.id}
```
```suggestion
      <strong>Layer ID:</strong> {feature.layer.id}
```

tristen commented on react-tooltip/src/components/Tooltip.js:
```diff
@@ -0,0 +1,17 @@
+
+const Tooltip = ({ feature }) => {
+  const { id } = feature.properties;
+  console.log(id);
```
Can you drop this?

tristen commented on react-tooltip/public/index.html:
```diff
@@ -1,14 +1,13 @@
-  <title></title>
-  <meta charset='utf-8'>
-  <meta name='viewport' content='width=device-width, initial-scale=1'>
-  <link href='https://api.mapbox.com/mapbox-assembly/mbx/v0.18.0/assembly.min.css' rel='stylesheet'>
-  <link href='https://api.tiles.mapbox.com/mapbox-gl-js/v0.39.1/mapbox-gl.css' rel='stylesheet' />
```
For all the examples, can we keep the references to these scripts?

tristen commented on react-tooltip/public/index.html:
```diff
@@ -1,14 +1,13 @@
-</head>
-<body>
-  <div id='app'></div>
-  <script src='https://api.mapbox.com/mapbox-assembly/mbx/v0.18.0/assembly.js'></script>
```
For all the examples, can we keep the reference to this script?

tristen commented on react-tooltip/README.md:
```diff
@@ -1,5 +1,4 @@
 React powered tooltip
----
```
Can we keep this? It's markdown to turn "React powered tooltip" into a heading

tristen commented on data-overlay/README.md:
```diff
@@ -1,5 +1,4 @@
 Data overlay example
----
```
Can we keep this? It's markdown to turn "Data overlay example" into a heading

tristen commented on data-overlay-redux/src/Map.js:
```diff
@@ -0,0 +1,107 @@
+      });
+
+      map.setLayoutProperty("country-label", "text-field", [
+        "format",
+        ["get", "name_en"],
+        { "font-scale": 1.2 },
+        "\n",
+        {},
+        ["get", "name"],
+        {
+          "font-scale": 0.8,
+          "text-font": [
+            "literal",
+            ["DIN Offc Pro Italic", "Arial Unicode MS Regular"],
+          ],
+        },
```
Nice addition!

tristen reviewed (changes requested):
Thanks so much for the changes so far and your patience. I left a more through code review, then this should be good to go after those changes.

**One broad change to make.** Can you change [all strings to use single quotes](https://github.com/airbnb/javascript#strings--quotes)?

---

I'm also not seeing the controls to switch between data and the legend in the `data-overlay` and `data-overlay-react` examples:

<img width="1440" alt="Screen Shot 2020-08-27 at 10 30 20 AM" src="https://user-images.githubusercontent.com/61150/91457267-27becf00-e852-11ea-86b6-913803537b82.png">

astridx committed:
npm ci

astridx committed:
add heading in readme

astridx committed:
add assembly to index.html again

astridx committed:
delete console.log

astridx committed:
improve label in tooltip

astridx committed:
improve label in tooltip2

astridx committed:
improve tooltip3

astridx committed:
change moveend to move

astridx committed:
prettier

astridx committed:
redo changes in data.json by prettier

astridx commented:
@tristen I am glad that you look at it and improve it.
I hope I've found everything you suggested. Sorry about the not working examples. The controls were not displayed because I accidentally deleted the CSS file. But now I've tested everything again. All examples should work.

tristen reviewed (approved):
@astridx the latest changes look great! I'm going to go ahead and merge this and follow up on documentation in #12
